### ISSUE TITLE
Remove `valueAsDate` and `valueAsNumber` from `<sl-input>`

### ISSUE THREAD

**Issue by claviska** [shoelace-style/webawesome#103]
Through [a request](https://github.com/shoelace-style/shoelace/issues/570) and subsequent issues (#760, shoelace-style/shoelace#796, shoelace-style/shoelace#1302, shoelace-style/shoelace#1303, shoelace-style/shoelace#1386 / shoelace-style/shoelace#1399, shoelace-style/shoelace#1704), it's become clear that the native `<input>` element's [`valueAsDate`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement#valueasdate) and [`valueAsNumber`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement#valueasnumber) properties are extremely difficult to support in a custom element.

I believe the issues are mostly due to 1) custom elements needing to be defined before using and 2) Lit's batching of updates, which make it difficult to provide these properties synchronously.

I propose we stop trying to mimic native inputs and replace these properties with methods:

```js
//
// Date inputs
//
input.valueAsDate(); // returns a Date object
input.valueAsDate(stringOrDateObject); // sets the value from a string or date object

//
// Number inputs
//
input.valueAsNumber(); // returns a number
input.valueAsNumber(number); // sets the value from a number
```

This will solve the timing issues we're seeing and the only requirement from users is to use `customElements.whenDefined()` like they already do.

Another issue with these properties is how they were programmed to attempt to mimic the native inputs. They use getters and setters to attempt to synchronously update the underlying input, but this special behavior differs greatly from how other properties are handled and could lead to issues with Lit's scheduled updates.

Related issue: https://github.com/shoelace-style/shoelace/issues/1704

**Comment by break-stuff** [comment]
Would setting the `input` property with the `queryAsync` decorator help in this situation for the setters?

**Comment by claviska** [comment]
> Would setting the input property with the queryAsync decorator help in this situation for the setters?

I don't think so because the native methods are synchronous. If we're OK with breaking that contract (which I think we probably should at this point), then it would probably work fine. However, exposing properties of the same name that don't work the same way might be quite confusing, especially since it [mostly] worked that way in the past.

**Comment by break-stuff** [comment]
Oh, I agree. I was thinking you could do something like:

```ts
@queryAsync('input')
private input?: Promise<HTMLInputElement>;

public get valueAsnumber() {
  return Number(this.value);
}

public set valueAsNumber(value: number) {
  this.input.then(x => x.value = value);
}
```

**Comment by break-stuff** [comment]
Actually, looking at your code, you are binding the `value` property directly to the input, so wouldn't even need to do that. It could be simpler.

```ts
public get valueAsnumber() {
  return Number(this.value);
}

public set valueAsNumber(value: number) {
  this.value = value?.toString() || '';
}
```

**Comment by claviska** [comment]
I'm trying to recall why I switched from `this.value` to using the input directly. It had something to do with Lit's lifecycle and the value not being ready in certain cases. Possibly the initial attribute/property mapping.

**Comment by claviska** [comment]
@KonnorRogers has a good opinion about this:

> My personal opinion is just get rid of them. I didn't even know people used that API; I personally don't trust it. It's like undocumented casting and I have no clue what the browser is going to do. If you wanna cast, use a proper Date or Number caster

I also vote to remove these seldom-used properties to clean things up and reduce confusion.

**Comment by claviska** [comment]
Fixed in #122
